lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -7518,6 +7549,31 @@ GenericSignature *GenericSignatureBuilder::computeRequirementSignature(
            /*allowBuilderToMove=*/false);
 }
 
+OpaqueTypeArchetypeType *
+GenericSignatureBuilder::computeOpaqueArchetype(OpaqueTypeDecl *Decl,
+                                            SubstitutionMap Subs,
+                                            RecursiveTypeProperties Props,
+                                            AllocationArena Arena) && {
+  finalize(SourceLoc(), getGenericParams(),
+           /*allow concrete generic params*/ true);
+  
+  auto equivClass =
+            this->resolveEquivalenceClass(Decl->getUnderlyingInterfaceType(),
+                                  ArchetypeResolutionKind::CompleteWellFormed);
+  
+  Type superclass;
+  LayoutConstraint layout;
+  SmallVector<ProtocolDecl*, 4> protos;
+  equivClass->getArchetypeConstraints(protos, superclass, layout);
```

Comment:
Rather than calling `finalize` / `resolveEquivalenceClass` / `getArchetypeConstraints`, how about calling `computeGenericSignature()` and then `GenericSignature::getConformsTo()` / `::getSuperclassBound()` / `::getLayoutConstraint()` ? The `GenericSignature` is going to get created at some point anyway.

Reply:
I'm trying to avoid forming a GenericSignature here, because it will be invalid if there are bindings to generic parameters outside of the signature.